src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -333,40 +353,48 @@ private void InnerStartOperationReceiveFrom()
 
         internal unsafe SocketError DoOperationReceiveFrom(SafeCloseSocket handle, out SocketFlags flags)
         {
-            PrepareIOCPOperation();
-
             flags = _socketFlags;
 
-            int bytesTransferred;
-            SocketError socketError;
-            if (_buffer != null)
+            SocketError socketError = SocketError.Success;
+            NativeOverlapped* overlapped = AllocateNativeOverlapped();
+            try
             {
-                socketError = Interop.Winsock.WSARecvFrom(
-                    handle,
-                    ref _wsaBuffer,
-                    1,
-                    out bytesTransferred,
-                    ref flags,
-                    _ptrSocketAddressBuffer,
-                    _ptrSocketAddressBufferSize,
-                    _ptrNativeOverlapped,
-                    IntPtr.Zero);
+                int bytesTransferred;
+
+                if (_buffer != null)
+                {
+                    socketError = Interop.Winsock.WSARecvFrom(
+                        handle,
+                        ref _wsaBuffer,
+                        1,
+                        out bytesTransferred,
+                        ref flags,
+                        _ptrSocketAddressBuffer,
+                        _ptrSocketAddressBufferSize,
+                        overlapped,
+                        IntPtr.Zero);
+                }
+                else
+                {
+                    socketError = Interop.Winsock.WSARecvFrom(
+                        handle,
+                        _wsaBufferArray,
+                        _bufferListInternal.Count,
+                        out bytesTransferred,
+                        ref flags,
+                        _ptrSocketAddressBuffer,
+                        _ptrSocketAddressBufferSize,
+                        overlapped,
+                        IntPtr.Zero);
+                }
+
+                socketError = ProcessIOCPResult(socketError == SocketError.Success, bytesTransferred);
+                return socketError;
             }
-            else
+            finally
             {
-                socketError = Interop.Winsock.WSARecvFrom(
-                    handle,
-                    _wsaBufferArray,
-                    _bufferListInternal.Count,
-                    out bytesTransferred,
-                    ref flags,
-                    _ptrSocketAddressBuffer,
-                    _ptrSocketAddressBufferSize,
-                    _ptrNativeOverlapped,
-                    IntPtr.Zero);
+                FreeNativeOverlappedIfNotPending(overlapped, socketError);
```

Comment:
Yeah, makes sense.

There's a lot more that happens in ProcessIOCPResult for SAEA than happens in ProcessOverlappedResult for IAsyncResult.  The intent is that the latter never throws, though looking at it now, it's not clear to me that there aren't some weird edge cases where it could throw...

Regardless:
(1) This code looks fine
(2) Longer term I think we should strive to have the minimal amount of code inside the try/catch -- i.e. just invoke the native call and process the result error code.  
(3) The only reason we have try/catch at all here is if the SafeCloseSocket marshalling throws -- if we ever revisit how this is handled we may be able to get rid of the try/catch entirely